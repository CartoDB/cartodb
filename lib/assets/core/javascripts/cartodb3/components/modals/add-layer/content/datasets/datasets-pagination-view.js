var CoreView = require('backbone/core-view');
var PaginationModel = require('../../../../pagination/pagination-model');
var PaginationView = require('../../../../pagination/pagination-view');

/**
 * Responsible for the datasets paginator
 *  ___________________________________________________________________________
 * |                                                                           |
 * |                                             Page 2 of 42 [1] 2 [3][4][5]  |
 * |___________________________________________________________________________|
 *
 */

module.exports = CoreView.extend({
  className: 'DatasetsPaginator',

  initialize: function (opts) {
    if (!opts.routerModel) throw new TypeError('routerModel is required');
    if (!opts.tablesCollection) throw new TypeError('tablesCollection is required');

    this._routerModel = opts.routerModel;
    this._tablesCollection = opts.tablesCollection;

    this.model = new PaginationModel({
      current_page: this._routerModel.get('page')
    });

    this._initBinds();
    this._initViews();
  },

  render: function () {
    this.clearSubViews();
    this.$el.append(this.paginationView.render().el);
    return this;
  },

  _initBinds: function () {
    this.model.bind('change', this.render, this);
    this.model.bind('change:current_page', function () {
      this._routerModel.set('page', this.model.get('current_page'));
    }, this);
    this._tablesCollection.bind('sync', this._updatePaginationModelByCollection, this);
    this._routerModel.bind('change:page', this._updatePaginationModelByRouterModel, this);

    this.add_related_model(this._routerModel);
    this.add_related_model(this._tablesCollection);
    this.add_related_model(this.model);
  },

  _initViews: function () {
    this.paginationView = new PaginationView({
      model: this.model
    });
    this.addView(this.paginationView);
  },

  _updatePaginationModelByCollection: function () {
    this.model.set({
      per_page: this._tablesCollection.getDefaultParam('per_page'),
      total_count: this._tablesCollection.getTotalStat('total_entries')
    });
  },

  _updatePaginationModelByRouterModel: function () {
    this.model.set('current_page', this._routerModel.get('page'));
  }

});
