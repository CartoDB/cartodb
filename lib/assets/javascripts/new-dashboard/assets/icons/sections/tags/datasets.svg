<svg width="14" height="16" xmlns="http://www.w3.org/2000/svg">
  <path d="M7 .5C3.645.5.25 1.53.25 3.5v9c0 1.97 3.395 3 6.75 3s6.75-1.03 6.75-3v-9c0-1.97-3.395-3-6.75-3zM7 2c3.41 0 5.25 1.065 5.25 1.5C12.25 3.935 10.41 5 7 5S1.75 3.935 1.75 3.5C1.75 3.065 3.59 2 7 2zm0 12c-3.41 0-5.25-1.065-5.25-1.5v-1.05c1.302.692 3.28 1.05 5.25 1.05 1.97 0 3.948-.359 5.25-1.05v1.05c0 .435-1.84 1.5-5.25 1.5zm0-3c-3.41 0-5.25-1.065-5.25-1.5V8.45C3.052 9.142 5.03 9.5 7 9.5c1.97 0 3.948-.359 5.25-1.05V9.5c0 .435-1.84 1.5-5.25 1.5zm0-3C3.59 8 1.75 6.935 1.75 6.5V5.45C3.052 6.141 5.03 6.5 7 6.5c1.97 0 3.948-.359 5.25-1.05V6.5C12.25 6.935 10.41 8 7 8z" fill="#2E3C43" fill-rule="nonzero"/>
</svg>
