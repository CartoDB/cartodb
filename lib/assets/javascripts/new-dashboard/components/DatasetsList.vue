<template>
  <div class="container grid">
    <div class="full-width">
      <SectionTitle class="grid-cell" :showActionButton="!selectedDatasets.length" ref="headerContainer">
        <template slot="icon">
          <img src="../assets/icons/section-title/data.svg" width="18" height="20" />
        </template>

        <template slot="title">
          <VisualizationsTitle
            :defaultTitle="$t(`DataPage.header.title['${appliedFilter}']`)"
            :selectedItems="selectedDatasets.length"/>
        </template>

        <template slot="dropdownButton">
          <DatasetBulkActions
            :selectedDatasets="selectedDatasets"
            :areAllDatasetsSelected="areAllDatasetsSelected"
            v-if="hasBulkActions && selectedDatasets.length"
            @selectAll="selectAll"
            @deselectAll="deselectAll"></DatasetBulkActions>

          <SettingsDropdown
            section="datasets"
            v-if="!selectedDatasets.length"
            :filter="appliedFilter"
            :order="appliedOrder"
            :orderDirection="appliedOrderDirection"
            :metadata="datasetsMetadata"
            @filterChanged="applyFilter">
            <img svg-inline src="../assets/icons/common/filter.svg">
          </SettingsDropdown>
        </template>

        <template slot="actionButton" v-if="showCreateButton">
          <CreateButton visualizationType="dataset" :disabled="!canCreateDatasets">
            {{ $t(`DataPage.createDataset`) }}
          </CreateButton>
        </template>

        <template v-if="shouldShowLimitsWarning" slot="warning">
          <NotificationBadge type="warning" :has-margin='false'>
            <div class="warning" v-html="$t('DataPage.header.warning', { counter: `${datasetsCount}/${datasetsQuota}`, path: upgradeUrl })"></div>
          </NotificationBadge>
        </template>

      </SectionTitle>
    </div>

    <div class="grid-cell grid-cell--col12" v-if="initialState">
      <InitialState :title="$t(`DataPage.zeroCase.title`)">
        <template slot="icon">
          <img svg-inline src="../assets/icons/datasets/initialState.svg">
        </template>
        <template slot="description">
          <p class="text is-caption is-txtGrey" v-html="$t(`DataPage.zeroCase.description`)"></p>
        </template>
        <template slot="actionButton">
          <CreateButton visualizationType="dataset" :disabled="!canCreateDatasets">{{ $t(`DataPage.zeroCase.createDataset`) }}</CreateButton>
        </template>
      </InitialState>
    </div>

    <div
        v-if="shouldShowHeader"
        class="grid-cell grid-cell--noMargin grid-cell--col12 grid__head--sticky"
        :class="{
          'has-user-notification': isNotificationVisible,
          'in-home': isInHomePage,
          'no-secondary-navbar': !hasSecondaryNavbar
        }">

      <DatasetListHeader :order="appliedOrder" :orderDirection="appliedOrderDirection" @changeOrder="applyOrder"></DatasetListHeader>
    </div>

    <ul class="grid-cell grid-cell--col12" v-if="!isFetchingDatasets && currentEntriesCount > 0">
      <li v-for="dataset in datasets" :key="dataset.id" class="dataset-item">
        <DatasetCard
        :dataset="dataset"
        :isSelected="isDatasetSelected(dataset)"
        :selectMode="isSomeDatasetSelected"
        :canHover="canHoverCard"
        @toggleSelection="toggleSelected"
        @contentChanged="onContentChanged">
      </DatasetCard>
      </li>
    </ul>

    <div class="grid-cell grid-cell--col12">
      <EmptyState
        :text="emptyStateText"
        v-if="emptyState">
        <img svg-inline src="../assets/icons/common/compass.svg">
      </EmptyState>
    </div>

    <ul v-if="isFetchingDatasets" class="grid-cell grid-cell--col12">
      <li v-for="n in maxVisibleDatasets" :key="n" class="dataset-item">
        <DatasetCardFake></DatasetCardFake>
      </li>
    </ul>

  </div>
</template>

<script>
import { mapState, mapGetters } from 'vuex';
import DatasetCard from '../components/Dataset/DatasetCard';
import DatasetListHeader from '../components/Dataset/DatasetListHeader';
import DatasetCardFake from '../components/Dataset/DatasetCardFake';
import SettingsDropdown from '../components/Settings/Settings';
import SectionTitle from 'new-dashboard/components/SectionTitle';
import VisualizationsTitle from 'new-dashboard/components/VisualizationsTitle';
import NotificationBadge from 'new-dashboard/components/NotificationBadge';
import InitialState from 'new-dashboard/components/States/InitialState';
import EmptyState from 'new-dashboard/components/States/EmptyState';
import CreateButton from 'new-dashboard/components/CreateButton';
import DatasetBulkActions from 'new-dashboard/components/BulkActions/DatasetBulkActions.vue';
import { shiftClick } from 'new-dashboard/utils/shift-click.service.js';
import * as accounts from 'new-dashboard/core/constants/accounts';

export default {
  name: 'DatasetsList',
  props: {
    maxVisibleDatasets: {
      type: Number,
      default: 12
    },
    hasBulkActions: {
      type: Boolean,
      default: false
    },
    canHoverCard: {
      type: Boolean,
      default: false
    }
  },
  components: {
    CreateButton,
    SettingsDropdown,
    SectionTitle,
    VisualizationsTitle,
    NotificationBadge,
    DatasetCard,
    DatasetCardFake,
    InitialState,
    EmptyState,
    DatasetListHeader,
    DatasetBulkActions
  },
  data () {
    return {
      isScrollPastHeader: false,
      selectedDatasets: [],
      lastCheckedItem: null
    };
  },
  computed: {
    ...mapState({
      appliedFilter: state => state.datasets.filterType,
      appliedOrder: state => state.datasets.order,
      appliedOrderDirection: state => state.datasets.orderDirection,
      datasets: state => state.datasets.list,
      datasetsMetadata: state => state.datasets.metadata,
      isFetchingDatasets: state => state.datasets.isFetching,
      filterType: state => state.datasets.filterType,
      currentEntriesCount: state => state.datasets.metadata.total_entries,
      totalUserEntries: state => state.datasets.metadata.total_user_entries,
      totalShared: state => state.datasets.metadata.total_shared,
      isFirstTimeViewingDashboard: state => state.config.isFirstTimeViewingDashboard,
      upgradeUrl: state => state.config.upgrade_url,
      planAccountType: state => state.user.account_type
    }),
    ...mapGetters({
      datasetsCount: 'user/datasetsCount',
      datasetsQuota: 'user/datasetsQuota',
      isOutOfDatasetsQuota: 'user/isOutOfDatasetsQuota'
    }),
    canCreateDatasets () {
      return this.$store.getters['user/canCreateDatasets'];
    },
    areAllDatasetsSelected () {
      return Object.keys(this.datasets).length === this.selectedDatasets.length;
    },
    shouldShowHeader () {
      return !this.emptyState && !this.initialState;
    },
    showCreateButton () {
      return (this.totalUserEntries || !this.isFirstTimeViewingDashboard) && !this.selectedDatasets.length;
    },
    initialState () {
      return this.isFirstTimeViewingDashboard &&
        !this.hasSharedDatasets &&
        !this.isFetchingDatasets &&
        this.hasFilterApplied('mine') &&
        this.totalUserEntries <= 0;
    },
    emptyState () {
      return ((!this.isFirstTimeViewingDashboard || this.hasSharedDatasets) || this.isFirstTimeViewerAfterAction) &&
        !this.isFetchingDatasets &&
        !this.currentEntriesCount;
    },
    emptyStateText () {
      const route = this.$router.resolve({name: 'datasets', params: { filter: 'shared' }});
      return this.hasSharedDatasets ? this.$t('DataPage.emptyCase.onlyShared', { path: route.href }) : this.$t('DataPage.emptyCase.default', { path: route.href });
    },
    isFirstTimeViewerAfterAction () {
      // First time viewing dashboard but user has performed any action such as drag and dropping a dataset (no page refreshing)
      return this.isFirstTimeViewingDashboard && this.currentEntriesCount <= 0 && !this.hasFilterApplied('mine');
    },
    hasSharedDatasets () {
      return this.totalShared > 0;
    },
    isSomeDatasetSelected () {
      return this.selectedDatasets.length > 0;
    },
    shouldShowLimitsWarning () {
      return this.isOutOfDatasetsQuota;
    },
    isNotificationVisible () {
      return this.$store.getters['user/isNotificationVisible'];
    },
    isInHomePage () {
      return this.$router.currentRoute.name === 'home';
    },
    hasSecondaryNavbar () {
      return !accounts.accountsWithDataCatalogLimits.includes(this.planAccountType);
    }
  },
  methods: {
    fetchDatasets () {
      this.$store.dispatch('datasets/fetch');
    },
    applyFilter (filter) {
      this.$emit('applyFilter', filter);
    },
    applyOrder (orderParams) {
      this.$emit('applyOrder', orderParams);
    },
    toggleSelected ({ dataset, isSelected, event }) {
      if (this.selectedDatasets.length && event.shiftKey) {
        this.doShiftClick(dataset);
        return;
      }

      if (isSelected) {
        this.lastCheckedItem = dataset;
        this.selectedDatasets.push(dataset);
        return;
      }

      this.selectedDatasets = this.selectedDatasets.filter(selectedDataset => selectedDataset.id !== dataset.id);
    },
    doShiftClick (dataset) {
      const datasetsArray = [...Object.values(this.datasets)];
      this.selectedDatasets = shiftClick(datasetsArray, this.selectedDatasets, dataset, this.lastCheckedItem || dataset);
    },
    selectAll () {
      this.selectedDatasets = [...Object.values(this.$store.state.datasets.list)];
    },
    deselectAll () {
      this.selectedDatasets = [];
    },
    isDatasetSelected (dataset) {
      return this.selectedDatasets.some(selectedDataset => selectedDataset.id === dataset.id);
    },
    hasFilterApplied (filter) {
      return this.filterType === filter;
    },
    getHeaderContainer () {
      return this.$refs.headerContainer;
    },
    onContentChanged (type) {
      this.$emit('contentChanged', type);
    }
  },
  watch: {
    selectedDatasets () {
      this.$emit('selectionChange', this.selectedDatasets);
    },
    totalUserEntries () {
      this.$store.dispatch('user/updateTableCount', this.totalUserEntries);
    }
  }
};
</script>

<style scoped lang="scss">
@import 'new-dashboard/styles/variables';

.full-width {
  width: 100%;
}

.grid__head--sticky {
  top: $header__height + $subheader__height;

  &.in-home,
  &.no-secondary-navbar {
    top: $header__height;
  }

  &.has-user-notification {
    top: $header__height + $subheader__height + $notification-warning__height;

    &.in-home,
    &.no-secondary-navbar {
      top: $header__height + $notification-warning__height;
    }
  }
}

.pagination-element {
  margin-top: 64px;
}

.dataset-item {
  &:not(:last-child) {
    border-bottom: 1px solid $border-color;
  }
}

.empty-state {
  margin: 20vh 0 8vh;
}

.warning {
  white-space: nowrap;
}
</style>
