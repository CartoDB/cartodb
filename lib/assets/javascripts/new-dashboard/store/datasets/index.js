import CartoNode from 'carto-node';
import toObject from 'new-dashboard/utils/to-object';
import Filters, { defaultParams as filtersDefaultParams } from 'new-dashboard/core/filters';

const client = new CartoNode.AuthenticatedClient();

const DEFAULT_VALUES = {
  filter: 'mine',
  order: 'updated_at',
  orderDirection: 'desc'
};

const datasets = {
  namespaced: true,
  state: {
    isFetching: false,
    isFiltered: false,
    isErrored: false,
    error: {},
    filterType: DEFAULT_VALUES.filter,
    order: DEFAULT_VALUES.order,
    orderDirection: DEFAULT_VALUES.orderDirection,
    list: {},
    metadata: {},
    page: 1,
    numPages: 1,
    resultsPerPage: 1
  },
  mutations: {
    setRequestError (state, error) {
      state.isFetching = false;
      state.isErrored = true;
      state.error = error;
    },

    setDatasets (state, datasets) {
      state.list = toObject(datasets.visualizations, 'id');
      state.metadata = {
        total_entries: datasets.total_entries,
        total_likes: datasets.total_likes,
        total_shared: datasets.total_shared,
        total_locked: datasets.total_locked,
        total_user_entries: datasets.total_user_entries
      };

      state.isFetching = false;
    },

    setFetchingState (state) {
      state.isFetching = true;
      state.isErrored = false;
      state.error = {};
    },

    setDatasetAttributes (state, {datasetId, datasetAttributes}) {
      Object.assign(state.list[datasetId], datasetAttributes);
    },

    setPagination (state, page = 1) {
      state.page = page;
      state.numPages = Math.ceil(state.metadata.total_entries / filtersDefaultParams.per_page) || 1;
    },

    setFilterType (state, filterType = 'mine') {
      state.filterType = filterType;
    },

    setOrder (state, orderOptions) {
      state.order = orderOptions.order || DEFAULT_VALUES.order;
      state.orderDirection = orderOptions.direction || DEFAULT_VALUES.orderDirection;
    },

    setResultsPerPage (state, resultsPerPage) {
      state.resultsPerPage = resultsPerPage;
    },

    updateLike (state, {datasetId, liked}) {
      state.list[datasetId].liked = liked;
    },

    updateNumberLikes (state, {datasetId, likes}) {
      state.list[datasetId].likes = likes;
    }
  },
  actions: {
    fetchDatasets (context) {
      context.commit('setFetchingState');

      const params = {
        ...Filters[context.state.filterType],
        types: 'table',
        page: context.state.page,
        order: context.state.order,
        order_direction: context.state.orderDirection,
        per_page: context.state.resultsPerPage,
        with_dependent_visualizations: 10
      };

      client.getVisualization('',
        params,
        function (err, _, data) {
          if (err) {
            context.commit('setRequestError', err);
            return;
          }
          context.commit('setDatasets', data);
          context.commit('setPagination', context.state.page);
        });
    },

    filterDatasets (context, filter) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', filter);
      context.dispatch('fetchDatasets');
    },

    orderDatasets (context, orderOptions) {
      context.commit('setPagination', 1);
      context.commit('setOrder', orderOptions);
      context.dispatch('fetchDatasets');
    },

    like (context, dataset) {
      const currentLikeStatus = dataset.liked;
      context.commit('updateLike', { datasetId: dataset.id, liked: true });
      client.like(dataset.id,
        function (err, _, data) {
          if (err) {
            context.commit('updateLike', { datasetId: dataset.id, liked: currentLikeStatus });
            return;
          }
          context.commit('updateNumberLikes', { datasetId: dataset.id, likes: data.likes });
        }
      );
    },

    deleteLike (context, dataset) {
      const currentLikeStatus = dataset.liked;
      context.commit('updateLike', { datasetId: dataset.id, liked: false });
      client.deleteLike(dataset.id,
        function (err, _, data) {
          if (err) {
            context.commit('updateLike', { datasetId: dataset.id, liked: currentLikeStatus });
            return;
          }
          context.commit('updateNumberLikes', { datasetId: dataset.id, likes: data.likes });
        }
      );
    },

    updateDataset (context, datasetOptions) {
      context.commit('setDatasetAttributes', datasetOptions);
    },

    setResultsPerPage (context, perPage) {
      context.commit('setResultsPerPage', perPage);
    },

    setURLOptions (context, options) {
      context.commit('setPagination', options.page);
      context.commit('setFilterType', options.filter);
      context.commit('setOrder', { order: options.order, direction: options.order_direction });
      context.dispatch('fetchDatasets');
    },

    resetFilters (context) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', 'mine');
      context.commit('setResultsPerPage', 12);
      context.commit('setOrder', { order: DEFAULT_VALUES.order, direction: DEFAULT_VALUES.orderDirection });
    }
  }
};

export default datasets;
