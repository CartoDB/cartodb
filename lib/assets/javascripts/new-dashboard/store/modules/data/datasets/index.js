import Filters from 'new-dashboard/core/configuration/filters';

import * as VisualizationActions from '../../../actions/visualizations';
import * as VisualizationMutations from '../../../mutations/visualizations';

const DEFAULT_VALUES = {
  filter: 'mine',
  order: 'favorited,updated_at',
  orderDirection: 'desc,desc'
};

const datasets = {
  namespaced: true,
  state: {
    isFetching: false,
    isFiltered: false,
    isErrored: false,
    error: {},
    filterType: DEFAULT_VALUES.filter,
    order: DEFAULT_VALUES.order,
    orderDirection: DEFAULT_VALUES.orderDirection,
    list: {},
    metadata: {},
    page: 1,
    numPages: 1,
    resultsPerPage: 1
  },
  mutations: {
    setFetchingState: VisualizationMutations.setFetchingState,
    setFilterType: VisualizationMutations.setFilterType,
    setOrder: VisualizationMutations.setOrder(DEFAULT_VALUES),
    setPagination: VisualizationMutations.setPagination,
    setResultsPerPage: VisualizationMutations.setResultsPerPage,
    setRequestError: VisualizationMutations.setRequestError,
    setVisualizations: VisualizationMutations.setVisualizations,
    updateNumberLikes: VisualizationMutations.updateNumberLikes,
    updateVisualization: VisualizationMutations.updateVisualization
  },
  actions: {
    deleteLike: VisualizationActions.deleteLike,
    filter: VisualizationActions.filter,
    like: VisualizationActions.like,
    order: VisualizationActions.order,
    resetFilters: VisualizationActions.resetFilters(DEFAULT_VALUES),
    setResultsPerPage: VisualizationActions.setResultsPerPage,
    setURLOptions: VisualizationActions.setURLOptions,
    updateVisualization: VisualizationActions.updateVisualization,

    fetch (context) {
      const params = {
        ...Filters[context.state.filterType],
        types: 'table',
        page: context.state.page,
        order: context.state.order,
        order_direction: context.state.orderDirection,
        per_page: context.state.resultsPerPage,
        with_dependent_visualizations: 10
      };

      VisualizationActions.fetchVisualizations(context, params);
    }
  }
};

export default datasets;
