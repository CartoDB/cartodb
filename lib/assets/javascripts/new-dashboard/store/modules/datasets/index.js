import toObject from 'new-dashboard/utils/to-object';
import Filters, { defaultParams as filtersDefaultParams } from 'new-dashboard/core/configuration/filters';

import * as VisualizationActions from '../../actions/visualizations';
import * as VisualizationMutations from '../../mutations/visualizations';

const DEFAULT_VALUES = {
  filter: 'mine',
  order: 'updated_at',
  orderDirection: 'desc'
};

const datasets = {
  namespaced: true,
  state: {
    isFetching: false,
    isFiltered: false,
    isErrored: false,
    error: {},
    filterType: DEFAULT_VALUES.filter,
    order: DEFAULT_VALUES.order,
    orderDirection: DEFAULT_VALUES.orderDirection,
    list: {},
    metadata: {},
    page: 1,
    numPages: 1,
    resultsPerPage: 1
  },
  mutations: {
    setRequestError (state, error) {
      state.isFetching = false;
      state.isErrored = true;
      state.error = error;
    },

    setVisualizations (state, datasets) {
      state.list = toObject(datasets.visualizations, 'id');
      state.metadata = {
        total_entries: datasets.total_entries,
        total_likes: datasets.total_likes,
        total_shared: datasets.total_shared,
        total_locked: datasets.total_locked,
        total_user_entries: datasets.total_user_entries
      };

      state.isFetching = false;
    },

    setFetchingState (state) {
      state.isFetching = true;
      state.isErrored = false;
      state.error = {};
    },

    setPagination (state, page = 1) {
      state.page = page;
      state.numPages = Math.ceil(state.metadata.total_entries / filtersDefaultParams.per_page) || 1;
    },

    setFilterType (state, filterType = 'mine') {
      state.filterType = filterType;
    },

    setOrder (state, orderOptions) {
      state.order = orderOptions.order || DEFAULT_VALUES.order;
      state.orderDirection = orderOptions.direction || DEFAULT_VALUES.orderDirection;
    },

    setResultsPerPage (state, resultsPerPage) {
      state.resultsPerPage = resultsPerPage;
    },

    like: VisualizationMutations.updateLike,
    updateNumberLikes: VisualizationMutations.updateNumberLikes
  },
  actions: {
    fetch (context) {
      const params = {
        ...Filters[context.state.filterType],
        types: 'table',
        page: context.state.page,
        order: context.state.order,
        order_direction: context.state.orderDirection,
        per_page: context.state.resultsPerPage,
        with_dependent_visualizations: 10
      };

      VisualizationActions.fetchVisualizations(context, params);
    },

    filterDatasets (context, filter) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', filter);
      context.dispatch('fetch');
    },

    orderDatasets (context, orderOptions) {
      context.commit('setPagination', 1);
      context.commit('setOrder', orderOptions);
      context.dispatch('fetch');
    },

    like: VisualizationActions.like,
    deleteLike: VisualizationActions.deleteLike,
    updateVisualization: VisualizationActions.updateVisualization,

    setResultsPerPage (context, perPage) {
      context.commit('setResultsPerPage', perPage);
    },

    setURLOptions (context, options) {
      context.commit('setPagination', options.page);
      context.commit('setFilterType', options.filter);
      context.commit('setOrder', { order: options.order, direction: options.order_direction });
      context.dispatch('fetch');
    },

    resetFilters (context) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', 'mine');
      context.commit('setResultsPerPage', 12);
      context.commit('setOrder', { order: DEFAULT_VALUES.order, direction: DEFAULT_VALUES.orderDirection });
    }
  }
};

export default datasets;
