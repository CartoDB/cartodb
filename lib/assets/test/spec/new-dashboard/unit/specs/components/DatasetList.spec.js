import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuex from 'vuex';
import datasets from '../fixtures/datasets';
import DatasetsList from 'new-dashboard/components/DatasetsList';

jest.mock('new-dashboard/components/CreateButton', () => {});

const localVue = createLocalVue();
localVue.use(Vuex);

describe('DatasetsList.vue', () => {
  it('should render properly', () => {
    const store = createStore();
    const propsData = {
      hasBulkActions: true,
      canHoverCard: true
    };
    const datasetsList = createDatasetsList(store, propsData);

    expect(datasetsList).toMatchSnapshot();
  });

  it('should render initial state', () => {
    const store = createStore({
      datasets: {
        isFetchingDatasets: false,
        list: [],
        metadata: {
          total_user_entries: 0,
          total_shared: 0
        },
        filterType: 'mine'
      },
      config: {
        isFirstTimeViewingDashboard: true
      }
    });
    const datasetsList = createDatasetsList(store);

    expect(datasetsList).toMatchSnapshot();
  });

  it('should render empty state', () => {
    const store = createStore({
      datasets: {
        isFetchingDatasets: false,
        list: [],
        metadata: {
          total_entries: 0,
          total_user_entries: 2
        },
        filterType: 'mine'
      },
      config: {
        isFirstTimeViewingDashboard: false
      }
    });
    const datasetsList = createDatasetsList(store);

    expect(datasetsList).toMatchSnapshot();
  });

  describe('Methods', () => {
    it('applyOrder: should dispatch datasets/order with order options', () => {
      const store = createStore();
      const datasetsList = createDatasetsList(store);

      datasetsList.vm.applyOrder({ order: 'updated_at', direction: 'asc' });

      expect(datasetsList.emitted('applyOrder').length).toBe(1);
    });

    it('applyFilter: should dispatch datasets/filter with filter type', () => {
      const store = createStore();
      const datasetsList = createDatasetsList(store);

      datasetsList.vm.applyFilter('locked');

      expect(datasetsList.emitted('applyFilter').length).toBe(1);
    });

    it('fetchDatasets: should dispatch datasets/fetch', () => {
      const store = createStore();
      spyOn(store, 'dispatch');
      const datasetsList = createDatasetsList(store);

      datasetsList.vm.fetchDatasets();

      expect(store.dispatch).toHaveBeenCalledWith('datasets/fetch');
    });

    it('hasFilterApplied: should return true if applied filter is equal to the one passed', () => {
      const store = createStore();
      const datasetsList = createDatasetsList(store);

      expect(datasetsList.vm.hasFilterApplied('mine')).toBe(true);
    });

    it('toggleSelected: adds the dataset passed to selected datasets list', () => {
      const store = createStore();
      const propsData = {
        canHoverCard: true
      };
      const datasetsList = createDatasetsList(store, propsData);

      const dataset = datasets.visualizations[0];
      const event = {
        shiftKey: shiftKey => false
      };
      const isSelected = true;
      datasetsList.vm.toggleSelected({ dataset, isSelected, event });

      expect(datasetsList.vm.selectedDatasets[0]).toEqual(dataset);
      expect(datasetsList.vm.lastCheckedItem).toEqual(dataset);
    });

    it('selectAll: marks all datasets as selected', () => {
      const store = createStore();
      const datasetsList = createDatasetsList(store);
      datasetsList.setData({
        selectedDatasets: []
      });
      datasetsList.vm.selectAll();

      expect(datasetsList.vm.selectedDatasets).toEqual(datasets.visualizations);
    });

    it('isDatasetSelected: returns true if dataset is selected', () => {
      const dataset = datasets.visualizations[0];
      const store = createStore();
      const datasetsList = createDatasetsList(store);
      datasetsList.setData({
        selectedDatasets: [dataset]
      });
      expect(datasetsList.vm.isDatasetSelected(dataset)).toBe(true);
    });

    it('onContentChanged: should emit a contentChanged event when DatasetCard emits it', () => {
      const store = createStore();
      const datasetsList = createDatasetsList(store);

      datasetsList.vm.onContentChanged();

      expect(datasetsList.emitted('contentChanged')).toBeTruthy();
    });

    describe('showCreateButton', () => {
      it('should return false if there is any dataset selected', () => {
        const store = createStore({
          datasets: {
            isFetchingDatasets: false,
            list: datasets.visualizations,
            metadata: {
              total_user_entries: 0,
              total_entries: 0
            },
            filterType: 'mine',
            order: 'updated_at',
            orderDirection: 'desc'
          },
          config: {
            isFirstTimeViewingDashboard: false
          },
          user: {
            account_type: 'account_type'
          }
        });
        const datasetsList = createDatasetsList(store);
        const dataset = datasets.visualizations[0];
        datasetsList.setData({
          selectedDatasets: [dataset]
        });

        expect(datasetsList.vm.showCreateButton).toEqual(false);
      });

      it('should return true if there is any user entry', () => {
        const store = createStore({
          datasets: {
            isFetchingDatasets: false,
            list: datasets.visualizations,
            metadata: {
              total_user_entries: 2,
              total_entries: 2
            },
            filterType: 'mine',
            order: 'updated_at',
            orderDirection: 'desc'
          },
          config: {
            isFirstTimeViewingDashboard: true
          },
          user: {
            account_type: 'account_type'
          }
        });
        const datasetsList = createDatasetsList(store);

        expect(datasetsList.vm.showCreateButton).toEqual(true);
      });

      it('should return true if it is not the first time the user sees the dashboard', () => {
        const store = createStore({
          datasets: {
            isFetchingDatasets: false,
            list: datasets.visualizations,
            metadata: {
              total_user_entries: 0,
              total_entries: 0
            },
            filterType: 'mine',
            order: 'updated_at',
            orderDirection: 'desc'
          },
          config: {
            isFirstTimeViewingDashboard: false
          },
          user: {
            account_type: 'account_type'
          }
        });
        const datasetsList = createDatasetsList(store);

        expect(datasetsList.vm.showCreateButton).toEqual(true);
      });

      it('should return false if there are no user entries and it the first time in the dashboard', () => {
        const store = createStore({
          datasets: {
            isFetchingDatasets: false,
            list: datasets.visualizations,
            metadata: {
              total_user_entries: 0,
              total_entries: 0
            },
            filterType: 'mine',
            order: 'updated_at',
            orderDirection: 'desc'
          },
          config: {
            isFirstTimeViewingDashboard: true
          },
          user: {
            account_type: 'account_type'
          }
        });
        const datasetsList = createDatasetsList(store);

        expect(datasetsList.vm.showCreateButton).toEqual(false);
      });
    });
  });
});

function createStore (customStoreData) {
  const storeData = {
    datasets: {
      isFetchingDatasets: false,
      list: datasets.visualizations,
      metadata: {
        total_user_entries: 2,
        total_entries: 2
      },
      filterType: 'mine',
      order: 'updated_at',
      orderDirection: 'desc'
    },
    config: {
      isFirstTimeViewingDashboard: false
    },
    user: {
      account_type: 'account_type'
    }
  };

  const storeInstance = new Vuex.Store({
    state: customStoreData || storeData
  });

  return storeInstance;
}

function createDatasetsList (store, propsData) {
  return shallowMount(DatasetsList, {
    mocks: {
      $t: key => key,
      $router: {
        resolve: href => href,
        currentRoute: {
          name: 'testRoute'
        }
      }
    },
    store,
    localVue,
    propsData
  });
}
