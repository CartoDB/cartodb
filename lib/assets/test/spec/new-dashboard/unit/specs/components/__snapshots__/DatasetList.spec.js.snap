// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetsList.vue should render empty state 1`] = `
<div class="container grid">
  <div class="full-width">
    <sectiontitle-stub showactionbutton="true" class="grid-cell"><template></template> <template></template> <template></template> <template></template>
      <!---->
    </sectiontitle-stub>
  </div>
  <!---->
  <!---->
  <!---->
  <div class="grid-cell grid-cell--col12">
    <emptystate-stub text="DataPage.emptyCase.default"><img svg-inline="" src="../assets/icons/common/compass.svg"></emptystate-stub>
  </div>
  <!---->
</div>
`;

exports[`DatasetsList.vue should render initial state 1`] = `
<div class="container grid">
  <div class="full-width">
    <sectiontitle-stub showactionbutton="true" class="grid-cell"><template></template> <template></template> <template></template>
      <!---->
      <!---->
    </sectiontitle-stub>
  </div>
  <div class="grid-cell grid-cell--col12">
    <initialstate-stub title="DataPage.zeroCase.title"><template></template> <template></template> <template></template></initialstate-stub>
  </div>
  <!---->
  <!---->
  <div class="grid-cell grid-cell--col12">
    <!---->
  </div>
  <!---->
</div>
`;

exports[`DatasetsList.vue should render properly 1`] = `
<div class="container grid">
  <div class="full-width">
    <sectiontitle-stub showactionbutton="true" class="grid-cell"><template></template> <template></template> <template></template> <template></template>
      <!---->
    </sectiontitle-stub>
  </div>
  <!---->
  <div class="grid-cell grid-cell--noMargin grid-cell--col12 grid__head--sticky">
    <datasetlistheader-stub order="updated_at" orderdirection="desc" issortable="true"></datasetlistheader-stub>
  </div>
  <ul class="grid-cell grid-cell--col12">
    <li class="dataset-item">
      <datasetcard-stub dataset="[object Object]" canhover="true" storeactiontype="datasets"></datasetcard-stub>
    </li>
    <li class="dataset-item">
      <datasetcard-stub dataset="[object Object]" canhover="true" storeactiontype="datasets"></datasetcard-stub>
    </li>
  </ul>
  <div class="grid-cell grid-cell--col12">
    <!---->
  </div>
  <!---->
</div>
`;
